botocore/client.py

```diff
@@ -199,37 +176,59 @@ def _resolve_verify_value(self, provided_verify_value, scoped_config):
         elif scoped_config is not None:
             return scoped_config.get('ca_bundle')
 
+    def _determine_region_name(self, region_name, client_config):
+        # Determine what region the user provided either via the
+        # region_name argument or the client_config. Note that
+        # endpoint resolvers are responsible for enforcing whether
+        # or not a region is required.
+        if region_name is None:
+            if client_config and client_config.region_name is not None:
+                region_name = client_config.region_name
+        return region_name
+
+    def _resolve_endpoint(self, region_name, endpoint_config, endpoint_url,
```

Comment:
Question, one thing that have never liked about the client class related code is that it has a bunch of internal methods that are used to resolve endpoints. What would your thoughts be of having an abstraction on top of the endpoint resolver that resolves all of the defaults for these pieces of data required to correctly create the client. So this is how I would see it (with a better class name):
```py
class ClientDefaultResolver(object):
    def __init__(self, endpoint_resolver):
       self._endpoint_resolver = endpoint_resolver
   def get_default_values(self, service_model, region, is_secure):
       endpoint_config = self._endpoint_resolver.construct_endpoint(service_model.endpoint_prefix, region_name)
       # Move logic from and do it:
       # Client._resolve_endpoint
       # Endpoint.create_endpoint_url
       # Client._get_preferred_signature_version
       return {endpoint_url, region_name, signing_name, signature_version}
```
From there I would use this helper class to get some default values that can then be overriden by values passed in by the client's constructor within the client class.

The main reason I am suggesting this is because this class is getting really bloated and hard to follow. It would be nice to separate some of this logic that does not need to live in the client class, just needs to be used by it. Thoughts?